pep-9999.rst

```diff
@@ -0,0 +1,723 @@
+to other literal types. For example, the following is legal::
+
+    ReadOnlyMode         = Literal["r", "r+"]
+    WriteAndTruncateMode = Literal["w", "w+", "wt", "w+t"]
+    WriteNoTruncateMode  = Literal["r+", "r+t"]
+    AppendMode           = Literal["a", "a+", "at", "a+t"]
+
+    AllModes = Literal[ReadOnlyMode, WriteAndTruncateMode, WriteNoTruncateMode, AppendMode]
+
+This feature is again intended to help make using and reusing literal types
+more ergonomic.
+
+**Note:** As a consequence of the above rules, type checkers are also expected
+to support types that look like the following::
+
+    Literal[Literal[Literal[1, 2, 3], "foo"], 5, 5, 5, None]
+
+This should be exactly equivalent to the following type::
+
+    Literal[1, 2, 3, "foo", 5, None]
+
+...and also to the following type::
+
+    Optional[Literal[1, 2, 3, "foo", 5]]
+
+
+Illegal parameters for ``Literal`` at type check time
+-----------------------------------------------------
+
+The following parameters are intentionally disallowed by design:
+
+- Arbitrary expressions like ``Literal[3 + 4]`` or
+  ``Literal["foo".replace("o", "b")]``. 
+
+  - Rationale: Literal types are meant to be a
+    minimal extension to the PEP 484 typing ecosystem and requiring type
+    checkers to interpret potentially expressions inside types adds too
+    much complexity. Also see `Rejected or out-of-scope ideas`_.
+
+  - As a consequence, complex numbers like ``Literal[4 + 3j]`` and ``Literal[-4 + 2j]``
+    are also prohibited. For consistency, literals like ``Literal[4j]`` that contain
+    just a single complex number are also prohibited.
+
+  - The only exception to this rule is the unary ``-`` (minus) for ints: types
+    like ``Literal[-5]`` are *accepted*.
+
+-  Tuples containing valid literal types like ``Literal[(1, "foo", "bar")]``.
+   The user could always express this type as
+   ``Tuple[Literal[1], Literal["foo"], Literal["bar"]]`` instead. Also,
+   tuples are likely to be confused with the ``Literal[1, 2, 3]``
+   shortcut.
+
+-  Mutable literal data structures like dict literals, list literals, or
+   set literals: literals are always implicitly final and immutable. So,
+   ``Literal[{"a": "b", "c": "d"}]`` is illegal.
+
+-  Any other types: for example, ``Literal[MyTypedDict]``, or
+   ``Literal[some_object_instance]`` are illegal.
+   This includes typevars: if ``T`` is a typevar,  ``Literal[T]`` is
+   not allowed. Typevars can vary over only types, never over values.
+
+The following are provisionally disallowed for simplicity. We can consider allowing
+them on a case-by-case basis based on demand.
+
+-  Explicit byte strings: e.g. ``Literal[b'foo']``.
+
+-  Explicit unicode strings: e.g. ``Literal[u'foo']``.
+
+-  Floats: e.g. ``Literal[3.14]``. Note: if we do decide to allow
+   floats, we should likely disallow literal infinity and literal NaN.
+
+-  Any: e.g. ``Literal[Any]`` Note: the semantics of what exactly
```

Comment:
You might be right -- I went back and forth on this a few times, but ended up leaning towards disallowing this because I was uncomfortable with the idea of anybody trying to explicitly construct a ``Literal[Any]``. I was also worried somebody might try and interpret this type as meaning "any literal" instead of "any type" or something like that.

But you're right, being able handle the literal + enums + bad import problem more smoothly would be pretty nice.